File: packages/csharp/Tree.cs
namespace csharp_packages;
using System.Collections.Generic;
using System;

public class BSTNode<T> {
    public T Data;
    public BSTNode<T>? Left, Right;
    public BSTNode(T data) { Data = data; }
}

public class BST<T> where T : IComparable<T> {
    public BSTNode<T>? Root;
    public void Insert(T data) {
        Root = InsertRec(Root, data);
    }
    private BSTNode<T> InsertRec(BSTNode<T>? root, T data) {
        if (root == null) return new BSTNode<T>(data);
        if (data.CompareTo(root.Data) < 0)
            root.Left = InsertRec(root.Left, data);
        else if (data.CompareTo(root.Data) > 0)
            root.Right = InsertRec(root.Right, data);
        return root;
    }
}

public class NaryTreeNode<T> {
    public T Data;
    public List<NaryTreeNode<T>> Children = new();
    public NaryTreeNode(T data) { Data = data; }
}

public class NaryTree<T> {
    public NaryTreeNode<T>? Root;
}

public class TrieNode {
    public System.Collections.Generic.Dictionary<char, TrieNode> Children = new();
    public bool IsEndOfWord;
}

public class Trie {
    private readonly TrieNode _root = new();
    
    public void Insert(string word) {
        var curr = _root;
        foreach (char c in word) {
            if (!curr.Children.ContainsKey(c)) {
                curr.Children[c] = new TrieNode();
            }
            curr = curr.Children[c];
        }
        curr.IsEndOfWord = true;
    }
    
    public bool Search(string word) {
        var node = SearchPrefix(word);
        return node != null && node.IsEndOfWord;
    }
    
    public bool StartsWith(string prefix) {
        return SearchPrefix(prefix) != null;
    }
    
    private TrieNode? SearchPrefix(string prefix) {
        var curr = _root;
        foreach (char c in prefix) {
            if (!curr.Children.ContainsKey(c)) return null;
            curr = curr.Children[c];
        }
        return curr;
    }
}

public class AVLTreeNode<T> {
    public T Data;
    public AVLTreeNode<T>? Left, Right;
    public int Height = 1;
    public AVLTreeNode(T data) { Data = data; }
}

public class AVLTree<T> where T : System.IComparable<T> {
    public AVLTreeNode<T>? Root;
    
    public void Insert(T data) {
        Root = InsertRec(Root, data);
    }
    
    private AVLTreeNode<T> InsertRec(AVLTreeNode<T>? node, T data) {
        if (node == null) return new AVLTreeNode<T>(data);
        
        int cmp = data.CompareTo(node.Data);
        if (cmp < 0) node.Left = InsertRec(node.Left, data);
        else if (cmp > 0) node.Right = InsertRec(node.Right, data);
        else return node; 
        
        node.Height = 1 + System.Math.Max(GetHeight(node.Left), GetHeight(node.Right));
        int balance = GetBalance(node);
        
        if (balance > 1 && data.CompareTo(node.Left!.Data) < 0) return RightRotate(node);
        if (balance < -1 && data.CompareTo(node.Right!.Data) > 0) return LeftRotate(node);
        if (balance > 1 && data.CompareTo(node.Left!.Data) > 0) {
            node.Left = LeftRotate(node.Left);
            return RightRotate(node);
        }
        if (balance < -1 && data.CompareTo(node.Right!.Data) < 0) {
            node.Right = RightRotate(node.Right);
            return LeftRotate(node);
        }
        return node;
    }
    
    private int GetHeight(AVLTreeNode<T>? node) => node?.Height ?? 0;
    
    private int GetBalance(AVLTreeNode<T>? node) => node == null ? 0 : GetHeight(node.Left) - GetHeight(node.Right);
    
    private AVLTreeNode<T> RightRotate(AVLTreeNode<T> y) {
        var x = y.Left!;
        var T2 = x.Right;
        x.Right = y;
        y.Left = T2;
        y.Height = 1 + System.Math.Max(GetHeight(y.Left), GetHeight(y.Right));
        x.Height = 1 + System.Math.Max(GetHeight(x.Left), GetHeight(x.Right));
        return x;
    }
    
    private AVLTreeNode<T> LeftRotate(AVLTreeNode<T> x) {
        var y = x.Right!;
        var T2 = y.Left;
        y.Left = x;
        x.Right = T2;
        x.Height = 1 + System.Math.Max(GetHeight(x.Left), GetHeight(x.Right));
        y.Height = 1 + System.Math.Max(GetHeight(y.Left), GetHeight(y.Right));
        return y;
    }
}

File: packages/js/tree.js
class BSTNode {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BinarySearchTree {
    constructor() {
        this.root = null;
    }
    insert(data) {
        const newNode = new BSTNode(data);
        if (this.root === null) {
            this.root = newNode;
        } else {
            this.insertNode(this.root, newNode);
        }
    }
    insertNode(node, newNode) {
        if (newNode.data < node.data) {
            if (node.left === null) {
                node.left = newNode;
            } else {
                this.insertNode(node.left, newNode);
            }
        } else {
            if (node.right === null) {
                node.right = newNode;
            } else {
                this.insertNode(node.right, newNode);
            }
        }
    }
}

class TreeNode {
    constructor(data) {
        this.data = data;
        this.children = [];
    }
    addChild(node) {
        this.children.push(node);
    }
}

class TrieNode {
    constructor() {
        this.children = {};
        this.isEndOfWord = false;
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();
    }
    insert(word) {
        let current = this.root;
        for (let char of word) {
            if (!current.children[char]) {
                current.children[char] = new TrieNode();
            }
            current = current.children[char];
        }
        current.isEndOfWord = true;
    }
    search(word) {
        let current = this.root;
        for (let char of word) {
            if (!current.children[char]) return false;
            current = current.children[char];
        }
        return current.isEndOfWord;
    }
    startsWith(prefix) {
        let current = this.root;
        for (let char of prefix) {
            if (!current.children[char]) return false;
            current = current.children[char];
        }
        return true;
    }
}

class AVLNode {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
        this.height = 1;
    }
}

class AVLTree {
    constructor() {
        this.root = null;
    }
    getHeight(node) {
        if (!node) return 0;
        return node.height;
    }
    getBalance(node) {
        if (!node) return 0;
        return this.getHeight(node.left) - this.getHeight(node.right);
    }
    rightRotate(y) {
        let x = y.left;
        let T2 = x.right;
        x.right = y;
        y.left = T2;
        y.height = Math.max(this.getHeight(y.left), this.getHeight(y.right)) + 1;
        x.height = Math.max(this.getHeight(x.left), this.getHeight(x.right)) + 1;
        return x;
    }
    leftRotate(x) {
        let y = x.right;
        let T2 = y.left;
        y.left = x;
        x.right = T2;
        x.height = Math.max(this.getHeight(x.left), this.getHeight(x.right)) + 1;
        y.height = Math.max(this.getHeight(y.left), this.getHeight(y.right)) + 1;
        return y;
    }
    insert(node, data) {
        if (!node) return new AVLNode(data);
        if (data < node.data) {
            node.left = this.insert(node.left, data);
        } else if (data > node.data) {
            node.right = this.insert(node.right, data);
        } else {
            return node;
        }

        node.height = 1 + Math.max(this.getHeight(node.left), this.getHeight(node.right));
        let balance = this.getBalance(node);

        if (balance > 1 && data < node.left.data) return this.rightRotate(node);
        if (balance < -1 && data > node.right.data) return this.leftRotate(node);
        if (balance > 1 && data > node.left.data) {
            node.left = this.leftRotate(node.left);
            return this.rightRotate(node);
        }
        if (balance < -1 && data < node.right.data) {
            node.right = this.rightRotate(node.right);
            return this.leftRotate(node);
        }
        return node;
    }
    insertData(data) {
        this.root = this.insert(this.root, data);
    }
}

module.exports = { BinarySearchTree, TreeNode, Trie, AVLTree };
